dowhy/causal_refuters/add_unobserved_common_cause.py

```diff
@@ -197,3 +198,191 @@ def include_confounders_effect(self, new_data, kappa_t, kappa_y):
+
+        for i in observed_variables:
+            column1 = given_data[i]
+            column2 = given_data[self._outcome_name[0]]
+            correlation_y = column1.corr(column2)
+            if correlation_y>=max_correlation_with_y:
+                max_correlation_with_y = correlation_y
+                column3 = new_data[self._treatment_name[0]]
+                max_correlation_with_z = column1.corr(column3)
+
+        correlation_y_list = []
+        correlation_z_list = []
+        new_metric_simulated_list = []    
+        x_list = []
+
+        for i in range(0, 1000, 100):
+            c1 = math.sqrt(i)
+            c2 = c1
+            U = []
+            d_y = list(pd.Series(standardized_residuals_y))
+            d_z = list(pd.Series(standardized_residuals_z))
+            for j in range(len(d_z)):
+                simulated_variable_mean = c1*d_y[j]+c2*d_z[j]
+                simulated_variable_stddev = 0.01
+                U.append(np.random.normal(simulated_variable_mean, simulated_variable_stddev, 1))
+            U = np.array(U)
+            model = sm.OLS(U,X)
+            results = model.fit()
+            U = U.reshape(-1, )
+            final_U = U - results.fittedvalues.values
+            new_data['simulated'] = final_U
+            column1 = new_data['simulated']
+            column2 = given_data[self._outcome_name[0]]
+            correlation_y = column1.corr(column2)
+            correlation_y_list.append(correlation_y)
+
+            column3 = new_data[self._treatment_name[0]]
+            correlation_z = column1.corr(column3)
+            correlation_z_list.append(correlation_z)
+
+            new_metric_simulated = correlation_y*correlation_z
+            new_metric_simulated_list.append(new_metric_simulated)
+            new_metric_observed = max_correlation_with_y*max_correlation_with_z
+
+            x_list.append(i)
+
+        if convergence_threshold == None:
+            convergence_threshold = 0.1
+        index = 1
+        while index<len(correlation_y_list):
+            if (correlation_y_list[index]-correlation_y_list[index-1])<=convergence_threshold:
+                c_star_max = x_list[index]
+                break
+            index = index+1
+
+        x = [i for i in range(0, 1000, 100)]
+
+        plt.plot(x, correlation_y_list, marker='o', markerfacecolor='blue', markersize=8, color='skyblue', linewidth=2)
+        plt.plot( x, correlation_z_list, marker='o', markerfacecolor='red', markersize=8, color='red', linewidth=2)
+        plt.plot( x , new_metric_simulated_list, marker='o', markerfacecolor='olive', markersize=8, color='green', linewidth=2)
+        plt.savefig('newmetricplot.png')
+
+        c1_final = 0
+        c2_final = 0
+
+        ans_c1_c2 = 10
+        i = 0.05
+
+        threshold = c_star_max/0.05
+
+        while i<=threshold:
+            c2 = i
+            c1 = c_star_max/c2
+            U = []
+            d_y = list(pd.Series(standardized_residuals_y))
```

Comment:
the code from 305 to 318 is being repeated. Can you create a separate method, `generate_unobs_confounder_from_residual(d_y, d_z, X, c1, c2, )`

Reply:
Whenever there is repeated code, always a good idea to create a fn